store/cachekv/store.go

```diff
@@ -43,8 +46,17 @@ func (store *Store) Clone() types.CacheKVStore {
 
 // Restore restores the store cache to a given snapshot.
 func (store *Store) Restore(s types.CacheKVStore) {
+	store.mtx.Lock()
+	defer store.mtx.Unlock()
+
 	ss := s.(*Store)
+	ss.mtx.Lock()
+	defer ss.mtx.Unlock()
```

Comment:
This introduces a possible deadlock.

Reply:
You have to first take a snapshot of the provided CacheKVStore, and then set the receiver store's state to that snapshot.

```go
func (s *Store) Restore(other types.CacheKVStore) {
    state := other.TakeSnapshot() // does locking internally, returns copy of state
    s.ApplySnapshot(state) // does locking internally
}
```

Reply:
I made a `swapCache` helper method, I think it looks good now, thanks for suggestions.

Reply:
The swapCache method still permits data races, unfortunately.